pkg/controller/clone-controller.go

```diff
@@ -102,162 +185,172 @@ func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+		} else if len(podList) > 1 {
+			return nil, nil, errors.Errorf("multiple target pods found for clone PVC %s/%s", pvc.Namespace, pvc.Name)
+		}
+		targetPod = podList[0]
+	}
+	return sourcePod, targetPod, nil
+}
+
 // Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
 // processing this pvc, which would result in multiple pods for the same pvc.
 func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
-	cr, err := getCloneRequestPVC(pvc)
+	anno := map[string]string{}
+
+	// find cloning source and target Pods
+	sourcePod, targetPod, err := c.findClonePodsFromCache(pvc)
 	if err != nil {
 		return err
 	}
-
-	// check our existing pvc one more time to ensure we should be working on it
-	// and to help mitigate any race conditions. This time we get the latest pvc.
-	doCreate, pvc, err := checkClonePVC(c.clientset, pvc, true)
-	if err != nil { // maybe an intermittent api error
+	pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+	if err != nil {
 		return err
 	}
-	if !doCreate { // don't create pods but not an error
-		return nil // forget key; logging already done
+
+	// Pods must be controlled by this PVC
+	if sourcePod != nil && !metav1.IsControlledBy(sourcePod, pvc) {
+		return errors.Errorf("found pod %s/%s not owned by pvc %s/%s", sourcePod.Namespace, sourcePod.Name, pvc.Namespace, pvc.Name)
+	}
+	if targetPod != nil && !metav1.IsControlledBy(sourcePod, pvc) {
+		return errors.Errorf("found pod %s/%s not owned by pvc %s/%s", targetPod.Namespace, targetPod.Name, pvc.Namespace, pvc.Name)
 	}
+	
+	// expectations prevent us from creating multiple pods. An expectation forces
+	// us to observe a pod's creation in the cache.
+	needsSync := c.podExpectations.SatisfiedExpectations(pvcKey)
 
-	// update pvc with 'AnnCloningPods' annotation to indicate cloning is in process
-	anno := map[string]string{AnnCloningPods: "exist"}
-	var lab map[string]string
-	pvc, err = updatePVC(c.clientset, pvc, anno, lab)
-	if err != nil {
-		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
+	// make sure not to reprocess a PVC that has already completed successfully,
+	// even if the pod no longer exists
+	previousPhase, exists := pvc.ObjectMeta.Annotations[AnnClonePodPhase]
+	if exists && (previousPhase == string(v1.PodSucceeded)) {
+		needsSync = false
 	}
 
-	//create random string to be used for pod labeling and hostpath name
-	generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
-	//create the source pod
-	pod, err := CreateCloneSourcePod(c.clientset, c.cloneImage, c.verbose, c.pullPolicy, cr, pvc, generatedLabelStr)
-	if err != nil {
-		//TODO: remove annotation AnnCloningPods from pvc as pod failed to run
-		return err
+	if needsSync && (sourcePod == nil || targetPod == nil){
+		//create random string to be used for pod labeling and hostpath name
+		generatedLabelStr := util.RandAlphaNum(GENERATED_CLONING_LABEL_LEN)
```

Comment:
if this string has to be the same between both pods, then you'll need to ensure it's only generated once.  RIght now if src pod gets created and target fails, then a different generated string will be used for the target pod the next time the creation is attempted. 

Maybe you could just use the target PVC's UUID here as the "random" value?

Reply:
I know we talked about that but I I want to ask this question...
RestartPolicy is 'Never' right now. can this be changed in the future?

Reply:
i don't think we'd want to do that. Restart policy only makes sense for long standing applications. In our case, we'd likely want to know that the pod failed and make a decision whether or not to start another one in our controller

Reply:
Done

Reply:
HOWEVER!! in the unit test, when I create the 'testing' pods I'm using a random string, or use the created 'testing' pvc (testPvc1) uid.  when I execute 'make test', I get error that it's different than the real pvc uid. Is this how it works? how does it know it's different? I just tried something to verify this, and  added hardcoded string, in both the real code and the testing code, and this case passed ("something like 'blabla'). If so, how can we make the unit test work? It will never know the cloning pvc uid....

Reply:
i'm not sure what you're experiencing here.

just make sure in the test that when you generate the source pvc, that you give it a UUID and set the appropriate labels/annotations in the expected target/src pods.  it should just work.

Reply:
There are couple of problems, but lets solve the first  one first:
pod contains two annotations. Since the annotation field is a map (the anno order is not known) I get error that the pod that I create and the pod that the controller creates are not identical. You can see at the output that the annotations are the same in both pods but not in the same order. The import pod has only one annotation so  there is not issue. I deleted one annotation only for testing purposes, and it passed. So I do not know how to handle this.
I am going to push the updated files, you will be able to see the error in travis.

Reply:
I think I figured out. Initially I tried to use similar test to t he importer but I can't.
!reflect.DeepEqual does not know to compare maps for example, so I will have to do it in my own method. I think that I will write something to check only the necessary fields instead comparing the whole pod object.